nowcasting_api/database_fast.py

```diff
@@ -90,3 +101,197 @@ def get_forecast_values_all_compact(
+    """Filter by gsp id."""
+    if gsp_ids is not None:
+        query = query.filter(model.gsp_id.in_(gsp_ids))
+    else:
+        # dont get gps id 0
+        query = query.filter(model.gsp_id != 0)
+    return query
+
+
+def filter_start_and_end_datetime(
+    query: Query,
+    end_datetime_utc: datetime | None,
+    start_datetime_utc: datetime | None,
+    model=ForecastValueLatestSQL,
+):
+    """Filter by start and end datetime."""
+    if start_datetime_utc is not None:
+        query = query.filter(model.target_time >= start_datetime_utc)
+    if end_datetime_utc is not None:
+        query = query.filter(model.target_time <= end_datetime_utc)
+    return query
+
+
+def get_forecasts(
+    session: Session,
+    start_datetime_utc: datetime | None = None,
+    end_datetime_utc: datetime | None = None,
+    gsp_ids=None,
+) -> ManyForecasts:
+    """Get forecast from the database.
+
+    We get all the latest forecast values for the blend model.
+    We convert the sqlalchemy objects to ManyForecasts
+    Particular focus has been put on only get the data we need from the database.
+
+    This function
+    1. get model ids
+    2. get forecast values from the forecast_value_latest table.
+        We tried to only get the relevant data as needed, as this is normally a larger query
+    3. get the forecasts we need
+    4. Convert to Forecast objects
+    5. add forecast values to forecast objects. We normalize the power,
+        and do the adjuster for gsp_id=0
+    6. convert to ManyForecasts
+
+    :param session: database session
+    :param start_datetime_utc: start datetime
+    :param end_datetime_utc: end datetime
+    :param gsp_ids: list of gsp ids
+    :return: ManyForecasts object
+    """
+
+    # 1. get model ids
+    model_ids = session.query(MLModelSQL.id).filter(MLModelSQL.name == "blend").all()
+    model_ids = [model_id[0] for model_id in model_ids]
+
+    # 2. get forecast values from database
+    query = session.query(
+        ForecastValueLatestSQL.target_time,
+        ForecastValueLatestSQL.expected_power_generation_megawatts,
+        ForecastValueLatestSQL.gsp_id,
+        ForecastValueLatestSQL.adjust_mw,
+    )
+
+    # distinct on target_time
+    query = query.distinct(ForecastValueLatestSQL.gsp_id, ForecastValueLatestSQL.target_time)
+
+    # join with model table
+    query = query.filter(ForecastValueLatestSQL.model_id.in_(model_ids))
+
+    # filters
+    query = filter_start_and_end_datetime(
+        query=query, start_datetime_utc=start_datetime_utc, end_datetime_utc=end_datetime_utc
+    )
+    query = filter_gsp_id(query=query, gsp_ids=gsp_ids)
+
+    # order by target time and created utc desc
+    query = query.order_by(
+        ForecastValueLatestSQL.gsp_id,
+        ForecastValueLatestSQL.target_time,
+        ForecastValueLatestSQL.created_utc.desc(),
+    )
+
+    forecast_values = query.all()
+
+    # 3. get forecasts
```

Comment:
Sounds like this will speed things up a lot, but still feels like this could be one efficient query instead of two separate ones – but we can circle back afterwards!

Reply:
Yea, happy to circle back on it. I think its actually better in two separate queries, otherwise we end up with lots of duplicate data coming back from the server. 

Lets say there are 317 forecasts with 14 datapoints, each with 100 forecast values with 4 database. 
If we do it with one query we ended up with 317*100*(14+4) = 570,600 bits of data
If we do it with two queries, we ended up with 317*14 + 317*100*4 = 131,238 bits of data, 23% of the data compare to 1 query.

Reply:
Yeah that feels like a wrong kind of JOIN but maybe I'm missing something here – anyway, one for later for sure, all good for now.